Enable Wasm OMG JIT on Windows

iangrunert opened the pull request:
#### 172739a00a897f316cd376d399c11fecee4d7ee7
<pre>
Enable Wasm OMG JIT on Windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=278116">https://bugs.webkit.org/show_bug.cgi?id=278116</a>

Reviewed by Yusuke Suzuki.

Six test failures remaining on JSTests/wasm.yaml:

wasm.yaml/wasm/gc/array_new_fixed_long.js.wasm-eager
wasm.yaml/wasm/gc/array_new_fixed_long.js.wasm-no-cjit
wasm.yaml/wasm/gc/array_new_fixed_long.js.wasm-aggressive-inline
wasm.yaml/wasm/lowExecutableMemory/executable-memory-oom.js.default-wasm
wasm.yaml/wasm/lowExecutableMemory/exports-oom.js.default-wasm
wasm.yaml/wasm/lowExecutableMemory/imports-oom.js.default-wasm

* Source/JavaScriptCore/assembler/MacroAssembler.h:
* Source/JavaScriptCore/assembler/ProbeContext.cpp:
(JSC::Probe::executeJSCJITProbe):
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/UGPRPair.h:
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::buildEntryBufferForCatch):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: <a href="https://commits.webkit.org/282316@main">https://commits.webkit.org/282316@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70591a71da16f15bb7268bfe98ac560263eab170

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62723 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42079 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15318 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66743 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13327 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64843 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49765 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13611 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50616 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9214 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65792 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39141 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54346 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31295 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35850 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11669 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12203 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/55833 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57399 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12000 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68438 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/61966 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6669 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11662 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57930 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6700 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54393 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58117 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5585 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83729 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9458 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37899 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14741 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38979 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40090 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38721 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

iangrunert commented on Source/JavaScriptCore/runtime/UGPRPair.h:
```diff
@@ -38,7 +38,7 @@ struct UGPRPair {
     UCPURegister second;
 };
-static_assert(sizeof(UGPRPair) >= sizeof(void*) * 2, "UGPRPair should fit in two machine registers");
+static_assert(sizeof(UGPRPair) == sizeof(void*) * 2, "UGPRPair should fit in two machine registers");
```
This doesn't work on the Apple Watch build, which is surprising to me.

Constellation replied:
This does not work on Apple Watch since it is using ARM64_32 ABI. You can change `void*` to UCPURegister.

Constellation reviewed (changes requested):
Commented on Apple Watch build failure.

Constellation reviewed (approved):
r=me

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.h:
```diff
@@ -115,6 +115,9 @@ struct ThrownExceptionInfo {
     void* payload;
 };
+#if USE(JSVALUE64)
+static_assert(sizeof(ThrownExceptionInfo) == sizeof(void*) * 2, "ThrownExceptionInfo should fit in two machine registers");
```
Change this to UCPURegister too

iangrunert committed:
Enable Wasm OMG JIT on Windows
https://bugs.webkit.org/show_bug.cgi?id=278116

Reviewed by Yusuke Suzuki.

Six test failures remaining on JSTests/wasm.yaml:

wasm.yaml/wasm/gc/array_new_fixed_long.js.wasm-eager
wasm.yaml/wasm/gc/array_new_fixed_long.js.wasm-no-cjit
wasm.yaml/wasm/gc/array_new_fixed_long.js.wasm-aggressive-inline
wasm.yaml/wasm/lowExecutableMemory/executable-memory-oom.js.default-wasm
wasm.yaml/wasm/lowExecutableMemory/exports-oom.js.default-wasm
wasm.yaml/wasm/lowExecutableMemory/imports-oom.js.default-wasm

* Source/JavaScriptCore/assembler/MacroAssembler.h:
* Source/JavaScriptCore/assembler/ProbeContext.cpp:
(JSC::Probe::executeJSCJITProbe):
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/UGPRPair.h:
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::buildEntryBufferForCatch):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: https://commits.webkit.org/282316@main
